docs: add session key docs (#1357)
alchemyplatform/aa-sdk: site/pages/smart-contracts/modular-account-v2/session-keys/adding-session-keys.mdx

---
title: Adding Session Keys to your Modular Account V2
description: Adding Session Keys to your Modular Account V2
---

:::tip
Session Keys are currently an experimental feature in the SDK. We are actively working on simplifying the usage, please note that there could be breaking changes as we improve this feature.
:::

# Adding session keys

> Adding a session key is simple with Account Kit! To add a session key, you need to 1) decide what permissions you want to grant the new key and 2) call the [`installValidation`](/smart-contracts/modular-account-v2/session-keys/adding-session-keys#install-validation-method) method on the account. This method will send a single user operation that will add the session key with scoped permission to your smart contract account on-chain. You can then use that session key to sign transactions for your account within the defined permissions!

Reviewer: > call the [`installValidation`](/smart-contracts/modular-account-v2/session-keys/adding-session-keys#install-validation-method) method on the account

is the account in this case the owner account that we're installing the session key on?

Adding scoped permissions to keys will happen via permission modules that you can pass as configuration parameters on the `installValidation` method. Permissions can be combined to fit your use case (e.g. you can limit a session key to only be able to spend 10 USDC within the next 24 hours on behalf of your account).

## Adding a global session key (i.e. additional owner)

This example shows:

1. Adding a global session key to your Modular Account V2. This essentially gives the session key full control of your account. Functionally, this is how you can add another owner on your smart account.

> 2. Adding a session key that can only call 'execute' on the account. Functionally, this allows the session key to have full control of the account other than changing the underlying the account implementation.

Reviewer: ```suggestion
2. Adding a session key that can only call 'execute' on the account. Functionally, this allows the session key to have full control of the account other than changing the underlying account implementation.
```

import { generatePrivateKey } from "viem/accounts";
import { type SmartAccountSigner } from "@aa-sdk/core";

const client = (
  await createModularAccountV2Client({
    chain: sepolia,
    transport: alchemy({ apiKey: "your-api-key" }),
    signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey()),
  })
).extend(installValidationActions);

let sessionKeyEntityId = 1;
const ecdsaValidationModuleAddress =
  getDefaultSingleSignerValidationModuleAddress(client.chain);
const sessionKeySigner: SmartAccountSigner =
  LocalAccountSigner.mnemonicToAccountSigner("SESSION_KEY_MNEMONIC");

// 1. Adding a session key with full permissions
await client.installValidation({
  validationConfig: {
    moduleAddress: ecdsaValidationModuleAddress,
    entityId: sessionKeyEntityId,
    isGlobal: true,
    isSignatureValidation: true,
    isUserOpValidation: true,
  },
  selectors: [],
  installData: SingleSignerValidationModule.encodeOnInstallData({
    entityId: sessionKeyEntityId,
    signer: await sessionKeySigner.getAddress(), // Address of the session key
  }),
  hooks: [],
});

// 2. Adding a session key that can only call `execute` on the account
sessionKeyEntityId = 2;

> const executeSelector = "0xb61d27f6";

Reviewer 2: Could we change this to pull the ABI like so: https://github.com/alchemyplatform/aa-sdk/blob/9230c8f9c9af7a9a7d90538ef6ec42dcca1c39db/examples/ui-demo/src/hooks/useRecurringTransactions.ts#L210-L213

Rather than hard-coding the selector.

Also, it might be best practice to show adding both `execute` and `executeBatch`.

Reviewer 2: Note: I'm not sure we correctly export this ABI, we might need to check that.

await client.installValidation({
  validationConfig: {
    moduleAddress: ecdsaValidationModuleAddress,
    entityId: sessionKeyEntityId,
    isGlobal: false,
    isSignatureValidation: false,
    isUserOpValidation: true,
  },
  selectors: [executeSelector],
  installData: SingleSignerValidationModule.encodeOnInstallData({
    entityId: sessionKeyEntityId,
    signer: await sessionKeySigner.getAddress(), // Address of the session key
  }),
  hooks: [],
});
```

## Adding a session key with permissions

### Time range

Configuring a session key with a time range allows you to limit how long the session key is valid for (e.g. only allow this key to sign on my account for the next day). The Time Range Module is used to enforce time-based validation for User Operations (UOs) in the system. This example will show you how to add a session key that starts in a day and expires in two days.

Additional Notes

> - the interval is inclusive i.e. `[beginningOfInterval, endOfInterval]`

Reviewer: might be worth mentioning that these values are Unix timestamps

```ts twoslash
import { createModularAccountV2Client } from "@account-kit/smart-contracts";
import {
  HookType,
  installValidationActions,
  getDefaultSingleSignerValidationModuleAddress,
  SingleSignerValidationModule,
  getDefaultTimeRangeModuleAddress,
  TimeRangeModule,
} from "@account-kit/smart-contracts/experimental";
import { LocalAccountSigner } from "@aa-sdk/core";
import { sepolia, alchemy } from "@account-kit/infra";
import { generatePrivateKey } from "viem/accounts";
import { type SmartAccountSigner } from "@aa-sdk/core";

const client = (
  await createModularAccountV2Client({
    chain: sepolia,
    transport: alchemy({ apiKey: "your-api-key" }),
    signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey()),
  })
).extend(installValidationActions);

let sessionKeyEntityId = 1;
const ecdsaValidationModuleAddress =
  getDefaultSingleSignerValidationModuleAddress(client.chain);
const sessionKeySigner: SmartAccountSigner =
  LocalAccountSigner.mnemonicToAccountSigner("SESSION_KEY_MNEMONIC");

const hookEntityId = 0; // Make sure that the account does not have a hook with this entity id on the module yet

> const validAfter = Math.floor(Date.now() / 1000) + 86400; // validity starts 1 day from now

Reviewer 2: Should we set this to 0 so the example is immediately usable from copy-pasting?

// Adding a session key that starts in a day and expires in two days
await client.installValidation({
  validationConfig: {
    moduleAddress: ecdsaValidationModuleAddress,
    entityId: sessionKeyEntityId,
    isGlobal: true,
    isSignatureValidation: true,
    isUserOpValidation: true,
  },
  selectors: [],
  installData: SingleSignerValidationModule.encodeOnInstallData({
    entityId: sessionKeyEntityId,
    signer: await sessionKeySigner.getAddress(), // Address of the session key
  }),
  hooks: [
    {
      hookConfig: {
        address: getDefaultTimeRangeModuleAddress(client.chain),
        entityId: hookEntityId,
        hookType: HookType.VALIDATION, // fixed value
        hasPreHooks: true, // fixed value
        hasPostHooks: false, // fixed value
      },
      initData: TimeRangeModule.encodeOnInstallData({
        entityId: hookEntityId,
        validAfter,
        validUntil,
      }),
    },
  ],
});
```

### Paymaster guard

#### Purpose

This module provides the ability to limit a session key to only be able to use a specific single paymaster.

#### Additional Notes

- you MUST specify a paymaster when using this module

> - if the paymaster that is registered with this module decides to no longer sponsor your user operations, the account would need to send a non user operation transaction to uninstall the module.

Reviewer 2: This could be bypassed the the root owner sending a user op, too

> - if the paymaster that is registered with this module decides to no longer sponsor your user operations, the account would need to send a non user operation transaction to uninstall the module.

Reviewer: what's an example of a non user operation that the session key can send to uninstall the module?

Author: this would be directly calling the account from an EOA

---
title: Adding Session Keys to your Modular Account V2
description: Adding Session Keys to your Modular Account V2
---

:::tip
Session Keys are currently an experimental feature in the SDK. We are actively working on simplifying the usage, please note that there could be breaking changes as we improve this feature.
:::

# Adding session keys

> Adding session keys is done with the `installValidation` method, with many configuration options. Below are some example configurations with code snippets. To learn more about the configuration parameters, check out the [install validation](smart-contracts/modular-account-v2/session-keys/adding-session-keys.mdx#install-validation-method) section below.

Reviewer 3: ```suggestion
Adding a session key is simple with Account Kit! To add a session key, you need to 1) decide what permissions you want to grant the new key and 2) call the [`installValidation`](/smart-contracts/modular-account-v2/session-keys/adding-session-keys#install-validation-method) method on the account. This method will send a single user operation that will add the session key with scoped permission to your smart contract account on-chain. You can then use that session key to sign transactions for your account within the defined permissions!

Adding scoped permissions to keys will happen via permission modules that you can pass as configuration parameters on the `installValidation` method. Permissions can be combined to fit your use case (e.g. you can limit a session key to only be able to spend 10 USDC within the next 24 hours on behalf of your account).
```

> ## Adding a simple session key

Reviewer 3: ```suggestion
## Adding a global session key (i.e. additional owner)
```

> This example is how you would add a session key to your Modular Account V2.

Reviewer 3: ```suggestion
This example shows:

1. Adding a global session key to your Modular Account V2. This essentially gives the session key full control of your account. Functionally, this is how you can add another owner on your smart account.
2. Adding a session key that can only call 'execute' on the account. Functionally, this allows the session key to have full control of the account other than changing the underlying the account implementation.

```

```ts twoslash
import { createModularAccountV2Client } from "@account-kit/smart-contracts";
import {
  installValidationActions,
  getDefaultSingleSignerValidationModuleAddress,
  SingleSignerValidationModule,
} from "@account-kit/smart-contracts/experimental";
import { LocalAccountSigner } from "@aa-sdk/core";
import { sepolia, alchemy } from "@account-kit/infra";
import { generatePrivateKey } from "viem/accounts";
import { type SmartAccountSigner } from "@aa-sdk/core";

const client = (
  await createModularAccountV2Client({
    chain: sepolia,
    transport: alchemy({ apiKey: "your-api-key" }),
    signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey()),
  })
).extend(installValidationActions);

let sessionKeyEntityId = 1;
const ecdsaValidationModuleAddress =
  getDefaultSingleSignerValidationModuleAddress(client.chain);
const sessionKeySigner: SmartAccountSigner =
  LocalAccountSigner.mnemonicToAccountSigner("SESSION_KEY_MNEMONIC");

> // Adding a session key with full permissions

Reviewer 3: ```suggestion
// 1. Adding a session key with full permissions
```

await client.installValidation({
  validationConfig: {
    moduleAddress: ecdsaValidationModuleAddress,
    entityId: sessionKeyEntityId,
    isGlobal: true,
    isSignatureValidation: true,
    isUserOpValidation: true,
  },
  selectors: [],
  installData: SingleSignerValidationModule.encodeOnInstallData({
    entityId: sessionKeyEntityId,
    signer: await sessionKeySigner.getAddress(), // Address of the session key
  }),
  hooks: [],
});

> // Adding a session key that can only call `execute` on the account

Reviewer 3: ```suggestion
// 2. Adding a session key that can only call `execute` on the account
```

sessionKeyEntityId = 2;
const executeSelector = "0xb61d27f6";

await client.installValidation({
  validationConfig: {
    moduleAddress: ecdsaValidationModuleAddress,
    entityId: sessionKeyEntityId,
    isGlobal: false,
    isSignatureValidation: false,
    isUserOpValidation: true,
  },
  selectors: [executeSelector],
  installData: SingleSignerValidationModule.encodeOnInstallData({
    entityId: sessionKeyEntityId,
    signer: await sessionKeySigner.getAddress(), // Address of the session key
  }),
  hooks: [],
});
```

## Adding a session key with permissions

> ### Adding a session key with a time range

Reviewer 3: ```suggestion
### Time range
```

> #### Purpose

Reviewer 3: ```suggestion
```

> 

Reviewer 3: ```suggestion
```

> The Time Range Module is used to enforce time-based validation for User Operations (UOs) in the system. It ensures that operations are only valid within a specified time range.

Reviewer 3: ```suggestion
Configuring a session key with a time range allows you to limit how long the session key is valid for (e.g. only allow this key to sign on my account for the next day). The Time Range Module is used to enforce time-based validation for User Operations (UOs) in the system. This example will show you how to add a session key that starts in a day and expires in two days.
```

Reviewer 3: idk if this is true

> #### Additional Notes

Reviewer 3: ```suggestion
Additional Notes
```

const validAfter = Math.floor(Date.now() / 1000) + 86400; // validity starts 1 day from now
const validUntil = validAfter + 86400; // validity ends 2 days from now

// Adding a session key that starts in a day and expires in two days
await client.installValidation({
  validationConfig: {
    moduleAddress: ecdsaValidationModuleAddress,
    entityId: sessionKeyEntityId,
    isGlobal: true,
    isSignatureValidation: true,
    isUserOpValidation: true,
  },
  selectors: [],
  installData: SingleSignerValidationModule.encodeOnInstallData({
    entityId: sessionKeyEntityId,
    signer: await sessionKeySigner.getAddress(), // Address of the session key
  }),
  hooks: [
    {
      hookConfig: {
        address: getDefaultTimeRangeModuleAddress(client.chain),
        entityId: hookEntityId,
        hookType: HookType.VALIDATION, // fixed value
        hasPreHooks: true, // fixed value
        hasPostHooks: false, // fixed value
      },
      initData: TimeRangeModule.encodeOnInstallData({
        entityId: hookEntityId,
        validAfter,
        validUntil,
      }),
    },
  ],
});
```

> ### Adding a session key with a paymaster guard

Reviewer 3: ```suggestion
### Paymaster guard
```

#### Purpose

> This module provides the ability to force user operations from an account to only be signed by a single paymaster.

Reviewer 3: ```suggestion
This module provides the ability to limit a session key to only be able to use a specific single paymaster.
```

Reviewer 3: Maybe this isn't true? does this apply to the whole account?

Author: for the session key yep

#### Additional Notes

> - you MUST specify a paymaster when using this module

Reviewer 3: a paymaster address? should we add our paymaster address here as an example

Reviewer: the paymaster associated with the ui demo gas policy?

const hookEntityId = 0; // Make sure that the account does not have a hook with this entity id on the module yet
const paymasterAddress = "0xd8da6bf26964af9d7eed9e03e53415d37aa96045";

// Adding a session key that can only use the above paymaster for user operations
await client.installValidation({
  validationConfig: {
    moduleAddress: ecdsaValidationModuleAddress,
    entityId: sessionKeyEntityId,
    isGlobal: true,
    isSignatureValidation: true,
    isUserOpValidation: true,
  },
  selectors: [],
  installData: SingleSignerValidationModule.encodeOnInstallData({
    entityId: sessionKeyEntityId,
    signer: await sessionKeySigner.getAddress(), // Address of the session key
  }),
  hooks: [
    {
      hookConfig: {
        address: getDefaultPaymasterGuardModuleAddress(client.chain),
        entityId: hookEntityId,
        hookType: HookType.VALIDATION, // fixed value
        hasPreHooks: true, // fixed value
        hasPostHooks: false, // fixed value
      },
      initData: PaymasterGuardModule.encodeOnInstallData({
        entityId: hookEntityId,
        paymaster: paymasterAddress,
      }),
    },
  ],
});
```

> ### Adding a session key with a native token and/or gas limit

Reviewer 3: ```suggestion
### Native token and/or gas limit
```

> 

Reviewer 3: ```suggestion
```

> #### Purpose

Reviewer 3: ```suggestion
```

> This module provides native token spending limits for modular accounts by:

Reviewer 3: ```suggestion
This module provides native token spending limits for modular accounts. Below we will show an example of adding a session key that has a 1 eth native token spend limit. Functionally, this module is enabled by:
```

- Tracking and limiting total native token spending across transactions
- Monitoring both direct transfers and gas costs from UserOperations
- Supporting special paymaster configurations for complex gas payment scenarios

#### Token Limit Features

- Tracks native token spending across:
  - Direct transfers via `execute`
  - Batch transfers via `executeBatch`
  - Contract creation via `performCreate`
  - UserOperation gas costs (when applicable)
- Supports special paymaster configurations for:
  - Standard paymasters (gas costs don't count against limit)
  - Special paymasters (gas costs do count against limit)
- Maintains separate limits per entity ID

#### Gas Cost Tracking

For UserOperations, the module tracks:

- Pre-verification gas
- Verification gas
- Call gas
- Paymaster verification gas (for special paymasters)
- Paymaster post-op gas (for special paymasters)

#### Additional Notes

- The module must be installed with both validation and execution hooks, the validation hook track gas, whereas the execution hook tracks value

> - Special paymasters must be explicitly configured to count against limits (SDK WIP)

Reviewer 3: Hmm this is confusing? Is this something we allow or is this not usable? I'd remove SDK WIP since people dont know what that means

> - Limits can be updated dynamically after installation (SDK WIP)

Reviewer 3: remove SDK WIP and either replace with unsupported or coming soon

    entityId: sessionKeyEntityId,
    isGlobal: true,
    isSignatureValidation: true,
    isUserOpValidation: true,
  },
  selectors: [],
  installData: SingleSignerValidationModule.encodeOnInstallData({
    entityId: sessionKeyEntityId,
    signer: await sessionKeySigner.getAddress(), // Address of the session key
  }),
  hooks: [
    {
      hookConfig: {
        address: getDefaultNativeTokenLimitModuleAddress(client.chain),
        entityId: hookEntityId,
        hookType: HookType.VALIDATION, // fixed value
        hasPreHooks: true, // fixed value
        hasPostHooks: false, // fixed value
      },
      initData: NativeTokenLimitModule.encodeOnInstallData({
        entityId: hookEntityId,
        spendLimit: parseEther("1"),
      }),
    },
    {
      hookConfig: {
        address: getDefaultNativeTokenLimitModuleAddress(client.chain),
        entityId: hookEntityId,
        hookType: HookType.EXECUTION, // fixed value
        hasPreHooks: true, // fixed value
        hasPostHooks: false, // fixed value
      },
      initData: "0x", // no initdata required as the limit was set up in the above installation call
    },
  ],
});
```

> ### Adding a session key with an allowlist or an ERC20 token limit

Reviewer 3: ```suggestion
### Allowlist or an ERC20 token limit
```

> 

Reviewer 3: ```suggestion
```

> #### Purpose

Reviewer 3: ```suggestion
```

import { type SmartAccountSigner } from "@aa-sdk/core";

const client = (
  await createModularAccountV2Client({
    chain: sepolia,
    transport: alchemy({ apiKey: "your-api-key" }),
    signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey()),
  })
).extend(installValidationActions);

let sessionKeyEntityId = 1;
const ecdsaValidationModuleAddress =
  getDefaultSingleSignerValidationModuleAddress(client.chain);
const sessionKeySigner: SmartAccountSigner =
  LocalAccountSigner.mnemonicToAccountSigner("SESSION_KEY_MNEMONIC");

const hookEntityId = 0; // Make sure that the account does not have a hook with this entity id on the module yet
const allowlistInstallData = AllowlistModule.encodeOnInstallData({
  entityId: hookEntityId,
  inputs: [
    {
      target: "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
      hasSelectorAllowlist: false, // whether to limit the callable functions on call targets
      hasERC20SpendLimit: false, // If "target" is an ERC20 token with a spend limit
      erc20SpendLimit: parseEther("1"), // The spend limit to set, if relevant
      selectors: [], // The function selectors to allow, if relevant
    },
  ],
});

> // Adding a session key that has a 1 eth native token spend limit

Reviewer 3: wait isnt this the same as the other example?

Author: yup should be erc token, changed

const executeSelector = toFunctionSelector(
  getAbiItem({
    abi: semiModularAccountBytecodeAbi,
    name: "execute",
  })
);

const executeBatchSelector = toFunctionSelector(
  getAbiItem({
    abi: semiModularAccountBytecodeAbi,
    name: "executeBatch",
  })
);

await client.installValidation({
  validationConfig: {
    moduleAddress: ecdsaValidationModuleAddress,
    entityId: sessionKeyEntityId,
    isGlobal: false,
    isSignatureValidation: false,
    isUserOpValidation: true,
  },
  selectors: [executeSelector, executeBatchSelector],
  installData: SingleSignerValidationModule.encodeOnInstallData({
    entityId: sessionKeyEntityId,
    signer: await sessionKeySigner.getAddress(), // Address of the session key
  }),
  hooks: [],
});
```

## Adding a session key with permissions

### Time range

Configuring a session key with a time range allows you to limit how long the session key is valid for (e.g. only allow this key to sign on my account for the next day). The Time Range Module is used to enforce time-based validation for User Operations (UOs) in the system. This example will show you how to add a session key that starts in a day and expires in two days.

> Additional Notes

Reviewer: ```suggestion
#### Additional Notes
```

Reviewer: to stay consistent with the other "Additional Notes" subheadings for the other modules

- the interval is inclusive i.e. `[beginningOfInterval, endOfInterval]`
- the values `beginningOfInterval` and `endOfInterval` are unix timestamps with a maximum size of uint32

> - the timestamp specifying the end of the interval must be **strictly greater than** the end of the interval

Reviewer 4: is this a typo?
```suggestion
- the timestamp specifying the end of the interval must be **strictly greater than** the beginning of the interval
```

Author: yup, good catch

```ts twoslash
import { createModularAccountV2Client } from "@account-kit/smart-contracts";
import {
  HookType,
  installValidationActions,
  getDefaultSingleSignerValidationModuleAddress,
  SingleSignerValidationModule,
  getDefaultTimeRangeModuleAddress,
  TimeRangeModule,
} from "@account-kit/smart-contracts/experimental";
import { LocalAccountSigner } from "@aa-sdk/core";
import { sepolia, alchemy } from "@account-kit/infra";
import { generatePrivateKey } from "viem/accounts";
import { type SmartAccountSigner } from "@aa-sdk/core";

const client = (
  await createModularAccountV2Client({
    chain: sepolia,
    transport: alchemy({ apiKey: "your-api-key" }),
    signer: LocalAccountSigner.privateKeyToAccountSigner(generatePrivateKey()),
  })
).extend(installValidationActions);

let sessionKeyEntityId = 1;
const ecdsaValidationModuleAddress =
  getDefaultSingleSignerValidationModuleAddress(client.chain);
const sessionKeySigner: SmartAccountSigner =
  LocalAccountSigner.mnemonicToAccountSigner("SESSION_KEY_MNEMONIC");

const hookEntityId = 0; // Make sure that the account does not have a hook with this entity id on the module yet

> const validAfter = 0; // validity once added

Reviewer 4: The text above (and comment below) state that we would add a session key that begins in a day & expires in 2 days. Wouldn't that mean the `validAfter` should be `current timestamp + 43200`?

Author: ah yeah, we modified the example so this has to change. it should be `validAfter  + 86400*2`

        hasPostHooks: false, // fixed value
      },
      initData: "0x", // no initdata required as the limit was set up in the above installation call
    },
  ],
});
```

### Allowlist or an ERC20 token limit

This module provides two key security features for modular accounts:

- **Allowlisting** - Controls which addresses and functions can be called
- **ERC-20 Spend Limits** - Manages spending limits for ERC-20 tokens

#### Allowlist Features

- Can specify permissions for:
  - Specific addresses + specific functions
  - Specific addresses + all functions (wildcard)
  - All addresses + specific functions (wildcard)
- Only applies to execute and executeBatch functions
- Permission checks follow this order:
  - If wildcard address → Allow
  - If wildcard function → Allow
  - If specific address + specific function match → Allow
  - Otherwise → Revert

#### ERC-20 Spend Limit Features

- Only allows transfer and approve functions for tracked tokens
- Works with standard execution functions:
  - execute
  - executeWithRuntimeValidation
  - executeUserOp
  - executeBatch

#### Additional Notes

- Module must be installed/uninstalled on an entity ID basis

> - Uninstalling for one entity ID doesn't affect other entities

Reviewer: how do you uninstall the hook for just one entity ID? from my understanding you can't uninstall a hook without also uninstalling validation

Author: hmm, the structure here is: 1 entity maps to 1 validation and n hooks. uninstalling for 1 entity uninstalls its validations and hooks, but doesnt affect hooks and validations of other entities

`installData`: This is the installation data that is passed to the validation module on installation. Each module has their own encoding for this data, so you would need to use the helper functions provided by that module.

`hooks`: This is an array of hooks to be installed on the key. Each element in the array contains a hookConfig object as well as initData to pass to the hook module.

`hookConfig`: This is the hook configuration to be applied to the session key.

- `address`: This is the address of the hook module to be installed on the session key.
- `entityId`: This is a hook module entity id that is different from the validation entity id. The decoupling enables multiple hooks provided by the same module to be applied on the same key. The only restriction here is that the hook module entity id should not be an entity id that's currently in use for the account.
- `hookType`: This specifies which phase should the hook be applied on, either HookType.VALIDATION to be a validation hook, or HookType.EXECUTION to be a pre and/or post-execution hook.

> - `hasPreHooks`: This specifies if the hook should be a pre hook.

Reviewer: nit: something more clear like "specifies that this hook runs before the validation hook being installed", same for `hasPostHooks`
